File: nodeClasses/legacyNode.m
classdef legacyNode < radioNode     % most basic one
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % Defines a node with the default behavior of always using the same
    % randomly chosen channel.
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    properties
    end
    
    methods
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % Constructor
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        function obj = legacyNode(numChans,numSteps,txProb)   % notice probabilty, tx after choose
            obj.actions = zeros(1,numChans);
            obj.actions(randi(numChans)) = 1;   % X = randi(imax) returns a pseudorandom scalar integer between 1 and imax.
            obj.numActions = size(obj.actions,1); 
            obj.actionTally = zeros(1,numChans+1);
            obj.actionHist = zeros(numSteps,numChans);
            obj.actionHistInd = zeros(1,numSteps);
            obj.txProbability = txProb;
        end    
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % Determines an action from the node's possible actions
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        function action = getAction(obj,stepNum)
            
            if rand <= obj.txProbability
                action = obj.actions;   %  obj.actions = zeros(1,numChans);
            else
                action = zeros(1,length(obj.actions));
            end
            
            obj.actionHist(stepNum,:) = action;
            if ~sum(action)
                obj.actionHistInd(stepNum) = 0;
            else
                obj.actionHistInd(stepNum) = find(action == 1) + 1;
            end
            
            if ~sum(action)
                obj.actionTally(1) = obj.actionTally(1) + 1;
            else
                obj.actionTally(2:end) = obj.actionTally(2:end) + action;
            end
        end
    end
end

File: legacyNode.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  1 16:16:16 2018

@author: Jet
"""

import numpy as np
import random
from radioNode import radioNode

class legacyNode(radioNode):
    

    
    def __init__(self, numChans, numSteps, txProbability, legacyChanIndex):
        self.actions                  = np.zeros(numChans)   
        self.actions[ legacyChanIndex ] = 1
        self.numActions               = np.size(self.actions,0)  
        self.actionTally              = np.zeros(numChans+1)
        self.actionHist               = np.zeros((numSteps,numChans))
        self.actionHistInd            = np.zeros(numSteps)
        self.txProbability            = txProbability
    
    def getAction(self, stepNum):             
        
        if random.random()  <= self.txProbability:
            action = self.actions 
        else:
            action = np.zeros(len(self.actions))  
       
            
        self.actionHist[stepNum,:] = action;  
        if not np.sum(action):
            self.actionHistInd[stepNum] = 0
        else:
            self.actionHistInd[stepNum] = np.where(action == 1)[0] + 1    
          
            
        if not np.sum(action):
            self.actionTally[0] = self.actionTally[0] + 1
        else:
            self.actionTally[1:] = self.actionTally[1:] + action
        return action
          